### PR TITLE
Jettons modules for TON

### PR DESCRIPTION
**Pull request by jzethar**
This PR proposes adding Jettons modules to 3xpl. Jetton is a TON token that can be fungible or non fungible. The PR includes 2 new modules and 1 trait for TON modules.

### CODE REVIEW

File: Modules/Common/TONLikeNFJettonModule.php [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                            $token_info = $this->api_get_nft_info($transaction['messageIn'][0]['transfer']['token']);
+
+                            $events[] = [
+                                'transaction' => $transaction['hash'],
+                                'currency'    => $token_info['collection_address'],
+                                'address'     => ($transaction['messageIn'][0]['transfer']['from'] === "") ? 'the-void' : $transaction['messageIn'][0]['transfer']['from'],
+                                'sort_key'    => $sort_key++,
+                                'effect'      => '-1',
+                                'extra'       => $token_info['index'],
+                                'extra_indexed' => $transaction['messageIn'][0]['transfer']['token'],
```
Comment (Har01d) [inline Modules/Common/TONLikeNFJettonModule.php:121]: Could you please comment on what's contained in `extra` and `extra_indexed`?

If `extra` contains unique NFT id, the extra model should be set like this: `public ?ExtraDataModel $extra_data_model = ExtraDataModel::Identifier;`

File: Modules/Common/TONLikeJettonModule.php [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+    public ?bool $should_return_events = true;
+    public ?bool $should_return_currencies = true;
+    public ?bool $allow_empty_return_events = true;
+    public ?bool $allow_empty_return_currencies = true;
+
+    public ?bool $mempool_implemented = false; // Technically, this is possible
+    public ?bool $forking_implemented = true;
+
+    public ?ExtraDataModel $extra_data_model = ExtraDataModel::Default;
```
Comment (Har01d) [inline Modules/Common/TONLikeJettonModule.php:38]: This is not necessary as there's no `extra` field for this module.

File: Modules/Common/TONLikeNFJettonModule.php (hunk 1 above)
Comment (jzethar) [reply Modules/Common/TONLikeNFJettonModule.php:121]: In TON every NFT has own address, but it also has an own index. Unfortunately, contracts realizations do not always support getting NFT address by index or index can be something else, so for data consistency it's better to collect NFT address as well. However, the comment about `$extra_data_model` is right, will update this

**Review by jokerthedon81 — COMMENTED** [review]
Joker

**Pull request merged**